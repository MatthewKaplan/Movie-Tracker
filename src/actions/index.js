export const fetchPopularMovies = popularMoviess => ({
  type: "FETCH_POPULAR_MOVIES",
  favorited: false,
  popularMoviess
});

export const fetchPopularTv = popularShows => ({
  type: "FETCH_POPULAR_TV",
  favorited: false,
  popularShows
});

export const fetchUpcoming = upcomingResults => ({
  type: "FETCH_UPCOMING",
  favorited: false,
  upcomingResults
});

export const fetchNews = newsResults => ({
  type: "FETCH_NEWS",
  newsResults
});

export const fetchGenres = genreResults => ({
  type: "FETCH_GENRES",
  favorited: false,
  genreResults
});

export const fetchSearch = searchResults => ({
  type: "FETCH_SEARCH",
  favorited: false,
  searchResults
});

export const setUser = user => ({
  type: "SET_USER",
  user
});

export const createUser = user => ({
  type: "CREATE_USER",
  user
});

export const isLoggedIn = bool => ({
  type: "IS_LOGGED_IN",
  isLoggedIn: bool
});

export const favoritesList = movies => ({
  type: "FAVORITES_LIST",
  favorited: false,
  movies
});

export const currentMovie = movie => ({
  type: "CURRENT_MOVIE",
  movie
});

export const handleError = errorMessage => ({
  type: "HANDLE_ERROR",
  errorMessage
});
