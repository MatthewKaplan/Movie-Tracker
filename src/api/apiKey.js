export const apiKey = "api_key=bc13bb74ba2ec1d960ae15ea70d535e5";

export default apiKey;