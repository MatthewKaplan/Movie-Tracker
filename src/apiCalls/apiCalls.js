import { BASE_URL } from '../util/path';

const fetchData = url => {
	return fetch(url).then(response => {
		if (!response.ok) {
			throw new Error('Fetch failed');
		} else {
			return response.json();
		}
	});
};

const fetchUserData = url => {
	return fetch(url).then(response => {
		if (!response.ok) {
			throw new Error('Fetch failed');
		} else {
			return response.json();
		}
	});
};

const fetchPost = async (url, options) => {
	try {
		const response = await fetch(url, options);
		return response.json();
	} catch (e) {
		console.log('Unable to Add a new note', e);
	}
};

const fetchAddUser = async (name, email, password) => {
  const url = `${BASE_URL}/users`;
	const body = { email, name, password };
	const options = {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(body)
  };
  
	try {
		const response = await fetch(url, options);
		return response.json();
	} catch (e) {
		console.log('Unable to Add a new note', e);
	}
};

const fetchUserLogin = async (email, password) => {
  const url = `${BASE_URL}/users/login`;
  const body = {email, password};
  const options = {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(body)
  };
  
  try {
		const response = await fetch(url, options);
		return response.json();
	} catch (e) {
		console.log('Unable to Add a new note', e);
	}
};

const fetchFavoriteMovies = async (id) => {
  const url = `${BASE_URL}/movies/favorites/${id}`;

  try {
    const favorites = await fetch(url);
    return favorites.json();
  } catch (e) {
    console.log('Unable to Add a new note', e);
  }
}

export { fetchPost, fetchData, fetchUserData, fetchAddUser, fetchUserLogin, fetchFavoriteMovies };
