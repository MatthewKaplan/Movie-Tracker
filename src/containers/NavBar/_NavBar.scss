@import "../../variables.scss";

.nav-bar {
	display: flex;
	position: absolute;
	z-index: 3;
	left: 0;
	right: 0;
	justify-content: center;
	max-width: 1280px;
	margin: 0 auto;
}

.header {
	display: flex;
	justify-content: space-between;
	align-content: center;
	width: 90vw;
}

.nav {
	margin-top: 1.5rem;
	color: white;
}

.register {
	margin-left: 220px;
}

.login {
	text-align: center;
	margin-top: 0;
	width: 32%;
}

.logout-btn {
	border: 1px solid white;
	border-radius: 25px;
	font-size: 16px;
	padding: 0 12px 20px 12px;
	margin: 0px 0.5rem;
	color: white;
	background-color: rgba(0, 0, 0, 0.123);
	&:hover {
		cursor: pointer;
		color: black;
		background-color: white;
	}
}

.greeting {
	color: white;
	margin: 0px 0.5rem;
	text-transform: uppercase;
}

.navBar-btn {
	display: flex;
	height: 20px;
	margin: 0 0.5rem;
}

.header-title {
	text-align: center;
	color: $text-w;
	width: 32%;
	font-family: "Righteous", cursive;
	margin: 0.8rem 0;
}

.left-page-links {
	display: flex;
	justify-content: space-around;
	width: 40%;
}

.right-page-links {
	display: flex;
	justify-content: space-between;
	align-items: center;
	width: 40%;
}

.search-box {
	position: absolute;
	top: 50%;
	left: 70%;
	transform: translate(-50%, -50%);
	background-color: rgba(0, 0, 0, 0.068);
	height: 30px;
	border-radius: 40px;
	padding: 2px;
	border: 1px solid white;
}

.search-box:hover > .search-txt {
	width: 180px;
	padding: 0 6px;
}

.search-box:hover > .search-btn {
	background: white;
	color: black;
}

.search-btn {
	color: #e84118;
	float: right;
	width: 30px;
	height: 30px;
	border-radius: 50%;
	background-color: rgba(0, 0, 0, 0.068);
	display: flex;
	justify-content: center;
	align-items: center;
	transition: 0.4s;
	color: white;
	cursor: pointer;
}

.search-btn > i {
	font-size: 16px;
}

.search-txt {
	border: none;
	background: none;
	outline: none;
	float: left;
	padding: 0;
	color: white;
	font-size: 16px;
	transition: 0.4s;
	line-height: 30px;
	width: 0px;
	font-weight: bold;
}

@media screen and (max-width: 414px) {
	.nav-bar {
		display: none;
	}
}
