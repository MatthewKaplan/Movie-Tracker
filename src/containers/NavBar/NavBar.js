import React from 'react';
import './_NavBar.scss';
import { NavLink, Link } from 'react-router-dom';
import { fetchSearch, setUser, isLoggedIn } from '../../actions';
import { connect } from 'react-redux';
import { fetchData } from '../../apiCalls/apiCalls';
import { apiKey } from '../../api/apiKey';
import PropTypes from 'prop-types';
import MobileNav from '../MobileNav/MobileNav';

export class NavBar extends React.Component {
	state = { error: '', userSearch: '' };

	searchChangeHandler = event => {
		const searchTerm = event.target.value;
		this.setState({
			userSearch: searchTerm
		});
		fetchData(`https://api.themoviedb.org/3/search/movie?${apiKey}&query=${searchTerm}`)
			.then(response => this.props.fetchSearch(response.results))
			.catch(error => this.setState({ error: error }));
	};

	render () {
		const renderLogin = (
			<React.Fragment>
				<NavLink to="/login" className="nav register login">
					Register
				</NavLink>
				<NavLink to="/Register" data-test="login-btn" className="nav login">
					Sign in
				</NavLink>
			</React.Fragment>
		);

		const renderLogout = (
			<div className="navBar-btn">
				<p className="greeting">Hello, {this.props.user && this.props.user.name}.</p>
				<button
					className="logout-btn"
					data-test="logout-btn"
					onClick={() => {
						this.props.setUser({});
						this.props.isLoggedIn(false);
					}}>
					Sign out
				</button>
			</div>
		);

		return (
			<React.Fragment>
				<div className="nav-bar">
					<header className="header">
						<div className="left-page-links">
							<NavLink to="/Movie-Tracker" className="nav">
								Home
							</NavLink>
							<NavLink to="/Favorites" className="nav">
								My Favorites
							</NavLink>
							<NavLink to="/Movies" className="nav">
								Movies
							</NavLink>
							<NavLink to="/TV_Shows" className="nav">
								TV Shows
							</NavLink>
						</div>
						<h1 className="header-title">Movie-Tracker</h1>
						<div className="right-page-links">
							<Link to="/SearchResults">
								<form className="searchForm">
									<div className="search-box">
										<input
											type="text"
											name=""
											className="search-txt"
											onChange={this.searchChangeHandler}
											placeholder="Type to search"
											data-test="search-input"
										/>
										<div className="search-btn">
											<i className="fas fa-search" />
										</div>
									</div>
								</form>
							</Link>
							{this.props.user.name ? renderLogout : renderLogin}
						</div>
					</header>
				</div>
				<MobileNav />
			</React.Fragment>
		);
	}
}

NavBar.propTypes = {
	searchResults: PropTypes.array,
	fetchSearch: PropTypes.func,
	setUser: PropTypes.func,
	isLoggedIn: PropTypes.func
};

export const mapStateToProps = state => {
	return {
		searchResults: state.search,
		user: state.user
	};
};

export const mapDispatchToProps = dispatch => ({
	fetchSearch: searchResults => dispatch(fetchSearch(searchResults)),
	setUser: user => dispatch(setUser(user)),
	isLoggedIn: bool => dispatch(isLoggedIn(bool))
});

export default connect(mapStateToProps, mapDispatchToProps)(NavBar);
