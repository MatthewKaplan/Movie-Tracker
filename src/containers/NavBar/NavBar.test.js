import React from "react";
import { shallow } from "enzyme";
import { NavBar, mapStateToProps, mapDispatchToProps } from "./NavBar";
import * as actions from "../../actions/index";
import MockData from "../../assets/mockData";
import { apiKey } from "../../api/apiKey";
import { fetchData } from "../../apiCalls/apiCalls";

jest.mock("../../apiCalls/apiCalls.js");

let mockSearch = MockData.searchResults;
let mockUser = { id: 1, name: "Matthew" };
let mockIsLoggedIn = true;
const isLoggedInMock = jest.fn();
const mockSetUser = jest.fn();

describe("NavBar", () => {
  let wrapper, instance;
  fetchData.mockImplementation(() => Promise.resolve(1));

  beforeEach(() => {
    wrapper = shallow(
      <NavBar
        search={mockSearch}
        user={mockUser}
        setUser={mockSetUser}
        isLoggedIn={isLoggedInMock}
      />
    );
    instance = wrapper.instance();
  });

  afterEach(() => {
    fetchData.mockClear();
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have a default state", () => {
    expect(wrapper.state()).toEqual({
      error: "",
      userSearch: ""
    });
  });

  it("should update the value of userSearch in state as user types", () => {
    expect(wrapper.state("userSearch")).toEqual("");
    let searchInput = { target: { value: "The Big Green" } };
    wrapper.find("[data-test='search-input']").simulate("change", searchInput);
    expect(wrapper.state("userSearch")).toEqual("The Big Green");
  });

  it("should invoke 'fetchData' with the correct param when searchChangeHandler is called", () => {
    expect(wrapper.state("userSearch")).toEqual("");
    let searchInput = { target: { value: "The Big Green" } };
    wrapper.find("[data-test='search-input']").simulate("change", searchInput);
    expect(wrapper.state("userSearch")).toEqual("The Big Green");
    expect(fetchData).toHaveBeenCalledWith(
      "https://api.themoviedb.org/3/search/movie?" +
        apiKey +
        "&query=" +
        wrapper.state("userSearch")
    );
  });

  describe("logout button", () => {
    it("should invoke 'setUser' with the correct param when the user clicks the log out button", () => {
      wrapper.find("[data-test='logout-btn']").simulate("click");
      expect(mockSetUser).toHaveBeenCalledWith({});
    });

    it("should invoke 'isLoggedIn' with the correct param when the user clicks the log out button", () => {
      wrapper.find("[data-test='logout-btn']").simulate("click");
      expect(isLoggedInMock).toHaveBeenCalledWith(false);
    });
  });
});

describe("mapStateToProps", () => {
  it("should return an object", () => {
    const mockData = {
      searchResults: [],
      user: {}
    };
    const expected = {
      user: {}
    };

    const mockprops = mapStateToProps(mockData);
    expect(mockprops).toEqual(expected);
  });
});

describe("mapDispatchToProps", () => {
  it("should call dispatch for fetchSearch", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = actions.fetchSearch(mockSearch);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.fetchSearch(mockSearch);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it("should call dispatch for setUser", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = actions.setUser(mockUser);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setUser(mockUser);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it("should call dispatch for isLoggedIn", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = actions.isLoggedIn(mockIsLoggedIn);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.isLoggedIn(mockIsLoggedIn);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
