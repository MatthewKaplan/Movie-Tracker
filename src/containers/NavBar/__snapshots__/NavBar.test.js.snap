// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogIn button should render the Login button 1`] = `ShallowWrapper {}`;

exports[`NavBar should match the snapshot 1`] = `ShallowWrapper {}`;
